### ISSUE TITLE
Put string constants for property names in namespaces and match names to values 

### ISSUE THREAD

**Issue by matty0ung** [Brewtarget/brewtarget#520]
I have been thinking more about the global constants ktableFoo, kpropBar, and so on.  At the moment they are a bit double-edged.  The benefit is that, if you refer to, say, `kpropAmountKg` in the code, you can be sure you've not amount_kg the "amount_kg" property name it refers to (which would be an embarrassing run-time error).  The downside is that, if you're looking for all the places we refer to the "amount_kg" property, then you have an extra layer of indirection.  To take a different example, if you want to find all the places we refer to equipment ID, you might end up doing the following:
```
$ grep -r equipment_id ../src
../src/database.cpp:      // Update equipment_id
../src/TableSchemaConst.h:static const QString kpropEquipmentId("equipment_id");
../src/TableSchemaConst.h:static const QString kcolEquipmentId("equipment_id");
../src/RecipeSchema.h:static const QString kcolRecipeEquipmentId("equipment_id");
$ grep -r kpropEquipmentId ../src
../src/database.cpp:   int id = get( tbl, parent->key(), tbl->foreignKeyToColumn(kpropEquipmentId)).toInt();
../src/database.cpp:                QString("%1=%2").arg(tbl->foreignKeyToColumn(kpropEquipmentId)).arg(newEquip->key()),
../src/TableSchemaConst.h:static const QString kpropEquipmentId("equipment_id");
../src/TableSchema.cpp:   m_foreignKeys[kpropEquipmentId] = new PropertySchema( kpropEquipmentId, kcolRecipeEquipmentId, QString("integer"), Brewtarget::EQUIPTABLE);
$ grep -r kcolEquipmentId ../src
../src/TableSchemaConst.h:static const QString kcolEquipmentId("equipment_id");
../src/TableSchema.cpp:         defineBtTable(kcolEquipmentId, Brewtarget::EQUIPTABLE);
$ grep -r kcolRecipeEquipmentId ../src
../src/RecipeSchema.h:static const QString kcolRecipeEquipmentId("equipment_id");
../src/TableSchema.cpp:   m_foreignKeys[kpropEquipmentId] = new PropertySchema( kpropEquipmentId, kcolRecipeEquipmentId, QString("integer"), Brewtarget::EQUIPTABLE);
$ 
```
It occurred to me that we might get the best of both worlds by using some namespaces, say Properties, Tables, Columns.  Then we could rename `kpropEquipmentId` to `Properties::equipment_id`, `ktableWater` to `Tables::water` and so on.  We retain the compiler-checking for typos but now make it easier to search for usages of a particular property/table/column.

**Comment by matty0ung** [comment]
From a quick count, we have about 450 definitions of ktableXxx kcolXxx and kpropXxx constants, and just under 1250 usages.  So the trick would be to script the change.  I'll put my bash/awk/sed/perl thinking hat on!

**Comment by matty0ung** [comment]
I got as far as writing a script that does most of the heavy lifting - including replacing string literals that aren't yet replaced by constants (eg in calls to `Brewtarget::mainWindow()->doOrRedoUpdate()`), but then I got to thinking about _where_ to define constants.

It feels to me like kpropXxxx constants belong in a different place from ktableXxxx and kcolXxxx.  The properties of an object are part of its public interface, available for use by the UI etc, but the DB table and column names are a private implementation detail that should be hidden from the UI.  Ideally, kpropXxxx or its equivalent should be defined in the same header file as the Q_PROPERTY to which it relates.  Even more ideally, it would be generated by the same macro as the property itself.  It looks like this is possible as follows:
```
//
// Q_PROPERTY_EXT extends the Q_PROPERTY macro so that the name of the property being defined is also available as a
// compile-time constant.
//
// Eg if used as follows:
//    class Foo : public QObject {
//       Q_PROPERTY_EXT(double, weight_kg, ...)
//       ...
//    }
// We can now refer to the weight_kg property by the following constant:
//    Foo::PROPERTY_weight_kg
// This gives advantages of having compile-time checking of some of our use of Qt's introspection.
//
// NB We have to use commas between parameters (INCLUDING between commented-out parameters) in Q_PROPERTY_EXT because
//    it is expanded by the compiler's preprocessor (which needs commas between parameters), whereas Q_PROPERTY is
//    expanded by the Qt Meta-Object Compiler (which doesn't accept commas between parameters, only whitespace).  So
//    we can't use a variadic macro.
//
#define Q_PROPERTY_EXT(type, name, p1, p2, p3, p4, p5, p6) \
   constexpr static char const * const PROPERTY_ ## name = "name"; \
   Q_PROPERTY(type name p1 p2 p3 p4 p5 p6)
```

Subject to what other folks think, I'll see if I can extend my script to make most of this sort of change programatically, but I'll probably wait until https://github.com/Brewtarget/brewtarget/pull/515 is merged before doing a pull request, as it will be another case of "the same change in a lot of places".

**Comment by matty0ung** [comment]
I played around with this some more, but it started looking rather ugly, so I've gone with something simpler.  For now I've focused just on the kprop globals.  The objectives were:
- Rename the property name constants so they match the names of the properties they refer to.  A property itself is just a name, so there's no inherent reason why we shouldn't use the same name for the corresponding constant, provided we avoid clashes with other names.  With namespaces this is possible. `kpropEvapRate` becomes `PropertyNames::Equipment::evapRate_pctHr`
- Move the property name constant definitions into the header files where the properties are defined.  If you want to use a property on a class, you probably already included that class's header for other reasons.  Equally you shouldn't need to include a header that has a bunch of "internal" stuff about database/XML column names etc if all you want is the name of a class's property.
- Fix places where we are using hard-coded strings to now refer to the property name constant, which gives us compile-time checking and improved readability (because, with the namespace versions of the constants, it's now clear which class we're referring to).  Eg `Brewtarget::setOption("colorMin_srm", invoked->data(),_section, Brewtarget::UNIT);` becomes `Brewtarget::setOption(PropertyNames::Style::colorMin_srm, invoked->data(),_section, Brewtarget::UNIT);`
- Leave in place the kprop constants that do not relate to real properties (eg kpropAncestorId, kpropHopId, kpropLocked,etc).  AFAICT most of these are used only by the database-related code.

I have a shell script that makes all the changes, so I can submit a patch now against head and then run the same script against https://github.com/Brewtarget/brewtarget/pull/515.
